File: AI/AIAssign1_DFS_BFS.py
from collections import defaultdict


class Graph:
    def __init__(self):
        self.graph = defaultdict(list)

    def add_edge(self, u, v):
        self.graph[u].append(v)
        self.graph[v].append(u)

    def dfs(self, v, visited):
        visited.add(v)
        print(v, end=" ")

        for neighbor in self.graph[v]:
            if neighbor not in visited:
                self.dfs(neighbor, visited)

    def bfs(self, v):
        visited = set()
        queue = []

        visited.add(v)
        queue.append(v)

        while queue:
            vertex = queue.pop(0)
            print(vertex, end=" ")

            for neighbor in self.graph[vertex]:
                if neighbor not in visited:
                    visited.add(neighbor)
                    queue.append(neighbor)


# Create a graph
graph = Graph()

# Take input from the user to add edges to the graph
num_edges = int(input("Enter the number of edges: "))
for _ in range(num_edges):
    u, v = input("Enter the edge (u v): ").split()
    graph.add_edge(u, v)

# Take input for the starting vertex
start_vertex = input("Enter the starting vertex: ")

# Perform DFS
print("Depth First Search (DFS):")
visited = set()
graph.dfs(start_vertex, visited)
print()

# Perform BFS
print("Breadth First Search (BFS):")
graph.bfs(start_vertex)
print()


'''
testcase:
Enter the number of edges: 5
Enter the edge (u v): A B
Enter the edge (u v): A C
Enter the edge (u v): B D
Enter the edge (u v): C E
Enter the edge (u v): D E
Enter the starting vertex: A
'''

File: AI/AIAssign1_DFS_BFS.cpp
#include <bits/stdc++.h>
using namespace std;

class Graph {
    unordered_map<char, unordered_set<char>> graph;

public:
    void addEdge(char u, char v) {
        graph[u].insert(v);
        graph[v].insert(u);
    }

    void dfs(char v, unordered_set<char>& visited) {
        visited.insert(v);
        cout << v << " ";

        for (char neighbor : graph[v]) {
            if (visited.find(neighbor) == visited.end()) {
                dfs(neighbor, visited);
            }
        }
    }

    void bfs(char v) {
        unordered_set<char> visited;
        queue<char> q;

        visited.insert(v);
        q.push(v);

        while (!q.empty()) {
            char vertex = q.front();
            q.pop();
            cout << vertex << " ";

            for (char neighbor : graph[vertex]) {
                if (visited.find(neighbor) == visited.end()) {
                    visited.insert(neighbor);
                    q.push(neighbor);
                }
            }
        }
    }
};

int main() {
    Graph graph;

    // Take input from the user to add edges to the graph
    int num_edges;
    cout << "Enter the number of edges: ";
    cin >> num_edges;

    cout << "Enter the edges (u v): " << endl;
    for (int i = 0; i < num_edges; i++) {
        char u, v;
        cin >> u >> v;
        graph.addEdge(u, v);
    }

    // Take input for the starting vertex
    char start_vertex;
    cout << "Enter the starting vertex: ";
    cin >> start_vertex;

    // Perform DFS
    cout << "Depth First Search (DFS):" << endl;
    unordered_set<char> visited;
    graph.dfs(start_vertex, visited);
    cout << endl;

    // Perform BFS
    cout << "Breadth First Search (BFS):" << endl;
    graph.bfs(start_vertex);
    cout << endl;

    return 0;
}




/*
testcase:
Enter the number of edges: 5
Enter the edge (u v): A B
Enter the edge (u v): A C
Enter the edge (u v): B D
Enter the edge (u v): C E
Enter the edge (u v): D E
Enter the starting vertex: A
*/